Cabal/src/Distribution/Types/LocalBuildConfig.hs

```diff
@@ -0,0 +1,187 @@
+    PackageBuildDescr (..)
+  , ComponentBuildDescr (..)
+  , LocalBuildDescr (..)
+  , LocalBuildConfig (..)
+  , BuildOptions (..)
+  ) where
+
+import Distribution.Compat.Prelude
+import Prelude ()
+
+import Distribution.Types.ComponentId
+import Distribution.Types.ComponentLocalBuildInfo
+import Distribution.Types.ComponentRequestedSpec
+import Distribution.Types.PackageDescription
+
+import Distribution.PackageDescription
+import Distribution.Simple.Compiler
+import Distribution.Simple.InstallDirs hiding
+  ( absoluteInstallDirs
+  , prefixRelativeInstallDirs
+  , substPathTemplate
+  )
+import Distribution.Simple.PackageIndex
+import Distribution.Simple.Program
+import Distribution.Simple.Setup.Config
+import Distribution.System
+
+import Distribution.Compat.Graph (Graph)
+
+-- | 'PackageBuildDescr' contains the information Cabal determines after
+-- performing package-wide configuration of a package, before doing any
+-- per-component configuration.
+data PackageBuildDescr = PackageBuildDescr
+  { configFlags :: ConfigFlags
+  -- ^ Options passed to the configuration step.
+  -- Needed to re-run configuration when .cabal is out of date
+  , flagAssignment :: FlagAssignment
+  -- ^ The final set of flags which were picked for this package
+  , componentEnabledSpec :: ComponentRequestedSpec
+  -- ^ What components were enabled during configuration, and why.
+  , compiler :: Compiler
+  -- ^ The compiler we're building with
+  , hostPlatform :: Platform
+  -- ^ The platform we're building for
+  , pkgDescrFile :: Maybe FilePath
+  -- ^ the filename containing the .cabal file, if available
+  , localPkgDescr :: PackageDescription
+  -- ^ WARNING WARNING WARNING Be VERY careful about using
+  -- this function; we haven't deprecated it but using it
+  -- could introduce subtle bugs related to
+  -- 'HookedBuildInfo'.
+  --
+  -- In principle, this is supposed to contain the
+  -- resolved package description, that does not contain
+  -- any conditionals.  However, it MAY NOT contain
+  -- the description with a 'HookedBuildInfo' applied
+  -- to it; see 'HookedBuildInfo' for the whole sordid saga.
+  -- As much as possible, Cabal library should avoid using
+  -- this parameter.
+  , installDirTemplates :: InstallDirTemplates
+  -- ^ The installation directories for the various different
+  -- kinds of files
+  -- TODO: inplaceDirTemplates :: InstallDirs FilePath
+  , withPackageDB :: PackageDBStack
+  -- ^ What package database to use, global\/user
+  }
+  deriving (Generic, Read, Show)
+
+-- | Information about individual components in a package,
+-- determined after the configure step.
+data ComponentBuildDescr = ComponentBuildDescr
+  { componentGraph :: Graph ComponentLocalBuildInfo
+  -- ^ All the components to build, ordered by topological
+  -- sort, and with their INTERNAL dependencies over the
+  -- intrapackage dependency graph.
+  -- TODO: this is assumed to be short; otherwise we want
+  -- some sort of ordered map.
```

Comment:
Is this still relevant? We do use an ordered map.

Reply:
I left all the comments on the constructors as-is, but I agree this is an opportunity to review them and update them. In particular I am hoping that all the complexity around the `localPkgDescr` field goes away as we get rid of `HookedBuildInfo` when we move away from `Custom`.

Reply:
I wish there was a way to make any progress on this situation, no matter how small.